File: leetcode/238-product-of-array-except-self.js
// https://leetcode.com/problems/product-of-array-except-self/ (top-interview-questions)

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
  const answer = new Array(nums.length).fill(1);

  let prefix = 1;
  for (let i = 0; i < nums.length; i += 1) {
    answer[i] = answer[i] * prefix;
    prefix *= nums[i];
  }

  let suffix = 1;
  for (let i = nums.length - 1; i >= 0; i -= 1) {
    answer[i] = answer[i] * suffix;
    suffix *= nums[i];
  }

  return answer;
};

File: leetcode/238-product-of-array-except-self.py
# https://leetcode.com/problems/product-of-array-except-self/ (top-interview-questions)

from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]: # Space: O(1)
        answer = [1] * len(nums)

        prefix = 1
        for i in range(len(nums)):
            answer[i] *= prefix
            prefix *= nums[i]
        
        suffix = 1
        for i in range(len(nums)-1, -1, -1):
            answer[i] *= suffix
            suffix *= nums[i]
        
        return answer


    def product_except_self(self, nums: List[int]) -> List[int]: # Space: O(n)
        prefix, suffix = [nums[0]], []
        for num in nums[1:]:
            prefix.append(prefix[-1] * num)
        
        for num in nums[::-1]:
            if not suffix:
                suffix.append(num)
                continue
            suffix.append(suffix[-1] * num)
        suffix = suffix[::-1]
        
        answer = [suffix[1]]
        for i in range(1, len(nums)-1):
            answer.append(prefix[i-1] * suffix[i+1])
        answer.append(prefix[-2])

        return answer
